Hourlymatching

wilson-ricks opened the pull request:
## Description


This PR updates the hourly matching policy to generalize it beyond hydrogen resources. Rather than creating a single "Qualified Supply" tag and an hourly matching constraint for each zone, it now allows for an arbitrary number of constraints (HM_1, HM_2, etc.) to be created. Demand can be added to each constraint in the following ways: via a pre-fixed time series in the Hourly_matching_requirement.csv input file, as a percentage of hourly demand in specified zones via the Hourly_matching_requirement_zonal file (similar to the current ESR policy), or as consumption from storage, flexible demand, or an electrolyzer by tagging these resources to the appropriate constraint in the Resource_hourly_matching_requirement.csv input file. Supply resources, flexible demand, and storage contribute to meeting the constraint if tagged in the same input file.

The hourly matching target (e.g. 0.9) is set via the first row entry in Hourly_matching_requirement.csv. Matching targets less than 1 are met via a new "shortfall" variable, which can contribute to the constraint in any hour but is limited in total annual quantity. A slack variable can also be created by including the Hourly_matching_requirement_slack.csv input file. 

Outputs include hourly matching prices, shortfall, and slack for each HM constraint.

The Hourly_matching_requirement.csv input file is updated to reflect the correct representative periods and included in time domain reduction folder if TDR is active, but hourly matching demand is not currently considered alongside actual demand, generation profiles, or fuel profiles when building representative periods.

Separately, I noticed an apparent bug that seemed to be only including energy revenues in the net revenue calculation for outputs. I believe I've corrected this so that all revenue categories will be included.


## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Bug Fix
- [x] Documentation Update
- [ ] Code Refactor
- [ ] Performance Improvements

## Related Tickets & Documents
n/a

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and .md files under /docs/src have been updated if necessary.
- [x] The latest changes on the target branch have been incorporated, so that any conflicts are taken care of before merging. This can be accomplished either by merging in the target branch (e.g. 'git merge develop') or by rebasing on top of the target branch (e.g. 'git rebase develop'). Please do not hesitate to reach out to the GenX development team if you need help with this.
- [x] Code has been tested to ensure all functionality works as intended.
- [unsure] CHANGELOG.md has been updated (if this is a 'notable' change).
- [x] I consent to the release of this PR's code under the GNU General Public license.

## How this can be tested

Try example systems 2 and 4, which have been updated to include hourly matching constraints and inputs. Suggest trying with the full range of includable technologies, including generation, storage, and flexible demand.

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  After the PR is approved we will give you a chance to tidy up the branch before merging.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 👷‍♀️ Create small PRs. In most cases, this will be possible. In the case of large feature or module additions, it is best to work towards a "minimum viable product" that is thoroughly tested but may not have all desired functionality or features, make a PR for this, and then later work to add more features.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation.

-->

## Post-approval checklist for GenX core developers
After the PR is approved

- [ ] Check that the latest changes on the target branch are incorporated, either via merge or rebase
- [ ] Remember to squash and merge if incorporating into develop

wilson-ricks committed:
Adds a comprehensive hourly matching policy module

This update replaces the existing "HydrogenHourlyMatching" policy and its "Qualified Supply" resource tag with a generalized hourly matching policy, where an unlimited number of constraints can be created matching demand (either fixed exogenously, set as a fraction of zonal demand like ESR policies, or based on electrolyzer consumption) is matched hourly with supply from tagged resources in each timestep. An optional slack variable is also available. Example systems 2 and 4 have been updated with the new policy.

Additionally, fixed an apparent bug where only energy revenue was being included in the resource net revenue calculation.

wilson-ricks committed:
Modified documentation

Modified documentation to align with the new hourly matching formulation.

wilson-ricks committed:
Bug fixes and documentation

Fixed bugs (references to deprecated settings tags) that popped up when merging.

Added further documentation and a change log entry.

wilson-ricks committed:
Documentation and test updates

Editing tests to reflect hourly matching updates

Adding correct documentation (hopefully)

wilson-ricks committed:
Test fix

Fixed error in electrolyzer test

wilson-ricks committed:
Fix_electrolyzer_test_final

Corrected the settings in the electrolyzer test to activate hourly matching

lbonaldo commented on src/model/generate_model.jl:
```diff
@@ -231,7 +231,7 @@ function generate_model(setup::Dict, inputs::Dict, OPTIMIZER::MOI.OptimizerWithA
 
     # Energy Share Requirement
-    if setup["HourlyMatching"] == 1
+    if setup["HourlyMatchingRequirement"] == 1
```
We need to fix the comment here

lbonaldo commented on src/write_outputs/write_outputs.jl:
```diff
@@ -365,14 +399,14 @@ function write_outputs(EP::Model, path::AbstractString, setup::Dict, inputs::Dic
-                println("Time elapsed for writing virtual discharge is")
-                println(elapsed_time_virtual_discharge)
-            end
+            #if output_settings_d["WriteVirtualDischarge"]
+            #    elapsed_time_virtual_discharge = @elapsed write_virtual_discharge(path,
+            #        inputs,
+            #        setup,
+            #        EP)
+            #    println("Time elapsed for writing virtual discharge is")
+            #    println(elapsed_time_virtual_discharge)
+            #end
```
is this intentional?

wilson-ricks replied:
No - not sure how I did that. Uncommented

lbonaldo commented on docs/src/Model_Reference/write_outputs.md:
````diff
@@ -121,6 +121,11 @@ Modules = [GenX]
 
+## Write Hourly Matching Revenue
+```@docs
+GenX.write_hourly_matching_revenue
````
I think we also need to include
```suggestion
GenX.write_hourly_matching_prices
GenX.write_hourly_matching_revenue
GenX.write_hourly_matching_slack
```

wilson-ricks replied:
Changed - was not sure about this one since other policy modules have output files that aren't listed here

lbonaldo commented on docs/src/User_Guide/slack_variables_overview.md:
```diff
@@ -35,6 +35,12 @@ Slack variables for Energy Share Requirement constraints are created when GenX d
+## Hourly Matching Requirement
+
+Slack variables for Hourly Matching Requirement constraints are created when GenX detects the presence of the file `Hourly_matching_requirement_slack.csv` in the Inputs folder. 
+This file should contain two columns: one titled 'HM_Constraint' naming the individual Energy Share Requirement constraints in the same order in which they are listed in the first row of `Hourly_matching_requirement.csv`, and a second titled 'PriceCap' containing the price thresholds for each constraint. 
```
```suggestion
This file should contain two columns: one titled 'HM_Constraint' naming the individual Hourly Matching Requirement constraints in the same order in which they are listed in the first row of `Hourly_matching_requirement.csv`, and a second titled 'PriceCap' containing the price thresholds for each constraint. 
```

lbonaldo commented on example_systems/2_three_zones_w_electrolyzer_and_hourly_matching/settings/genx_settings.yml:
```diff
@@ -1,9 +1,9 @@
 UCommit: 2 # Unit committment of thermal power plants; 0 = not active; 1 = active using integer clestering; 2 = active using linearized clustering
 StorageLosses: 1 # Energy Share Requirement and CO2 constraints account for energy lost; 0 = not active (DO NOT account for energy lost); 1 = active systemwide (DO account for energy lost)
-HydrogenHourlyMatching: 1 # Hydrogen electrolyzer contribution to hourly supply matching required
+#HydrogenHourlyMatching: 1 # Hydrogen electrolyzer contribution to hourly supply matching required
```
```suggestion
```

lbonaldo commented on src/load_inputs/load_hourly_matching_requirement.jl:
```diff
@@ -0,0 +1,49 @@
+        my_dir = path
+    end
+
+    filename = "Hourly_matching_requirement.csv"
+    hm_in = load_dataframe(joinpath(my_dir, filename))
+    inputs["HMCols"] = names(hm_in)[2:end]
+    # Absolute demand profiles and percentage matching targets for each hourly matching constriant
+    demand = Matrix(hm_in[2:end, 2:end]) ./ scale_factor # GWh if scaled, MWh if not scaled
+    matching_target = hm_in[1, 2:end] 
+    inputs["nHM"] = size(demand, 2)
+    inputs["dfHM_absolute"] = demand
+    inputs["dfHM_matching_target"] = matching_target
+    println(filename * " Successfully Read!")
```
Nice-to-have feature: to improve robustness, it would be nice here to mimic the approach used for loading `demand_data.csv` by using the `extract_matrix_from_dataframe` function in `load_demand_data!`. This would ensure that the column order in the final matrix is determined by column names (`HM_1`, `HM_2`, `HM_3`, etc) rather than their position in the table.

wilson-ricks replied:
Going to save this for another PR alongside the other load_inputs functions.

lbonaldo commented on src/load_inputs/load_resources_data.jl:
```diff
@@ -1294,25 +1260,14 @@ function add_resources_to_input_data!(inputs::Dict,
+    #         """The HydrogenHourlyMatching setting is deprecated. Please instead include electrolyzers in the relevant hourly matching constraint via the HM tag.""",
+    #         :add_resources_to_input_data!, force = true)
+    #     setup["HourlyMatching"] = 1
+    # end
```
```suggestion
```

lbonaldo commented on src/model/policies/hourly_matching.jl:
```diff
@@ -2,12 +2,15 @@
 
 This module defines the hourly matching policy constraint. 
-This constraint can be enabled by setting `HourlyMatching==1` in `genx_settings.yml`) requires generation from qualified resources ($y \in \mathcal{Qualified}$, indicated by `Qualified_Supply==1` in the `Resource_hourly_matching.csv` files) to be >= hourly consumption from electrolyzers in the zone and any charging by qualified storage within the zone used to help increase electrolyzer utilizatio ...
+This constraint can be enabled by setting `HourlyMatchingRequirement==1` in `genx_settings.yml`) requires generation from qualified resources ($y \in \mathcal{HM\_i}$, indicated by `HM_i==1` in the `Resource_hourly_matching.csv` file) to be >= hourly consumption from specified sources of demand.
```
```suggestion
This constraint can be enabled by setting `HourlyMatchingRequirement==1` in `genx_settings.yml`) requires generation from qualified resources ($y \in \mathcal{HM\_i}$, indicated by `HM_i==1` in the `Resource_hourly_matching_requirement.csv` file) to be >= hourly consumption from specified sources of demand.
```

lbonaldo commented on src/model/resources/flexible_demand/flexible_demand.jl:
```diff
@@ -84,6 +84,13 @@ function flexible_demand!(EP::Model, inputs::Dict, setup::Dict)
 
+    # Hourly matching policy
+    if setup["HourlyMatchingRequirement"] == 1
+        @expression(EP, eHMFlex[t = 1:T, HM = 1:inputs["nHM"]],
+            sum(hm(gen[y], tag = HM) * (EP[:vCHARGE_FLEX][y, t]-2*EP[:vP][y, t])
+            for y in intersect(ids_with_policy(gen, hm, tag = HM), FLEX)))
```
why `2*vP`?

wilson-ricks replied:
In discharge.jl I included an expression adding vP for every qualified resource to each hourly matching constraint. Rather than create a carveout for FLEX resources in that function, I decided to just subtract 2*vP here so we net out to the correct -vP. We should really change flexible_demand so vP contributes to the demand balance in the same direction as every other resource...

lbonaldo replied:
That makes sense, thank you!

lbonaldo reviewed (commented):
Hey @wilson-ricks, thanks so much for this addition! I did a first round of code review and left a few minor comments directly in the code. Let me know what you think! I still need to go through the changes in the TDR module.

One part I believe is missing is the update to the `model_input.md` page of the documentation. It would be good to describe the new input files as well as the new settings.

A final design suggestion: for the `Hourly_matching_requirement.csv` file, I would recommend splitting it into two separate files, one for the targets (which could be either YAML or CSV) and another for the time series data. This would improve clarity by making sure that each row in a file has a consistent meaning and structure. (In a similar way, we were thinking about updating the `Fuels_data.csv` file to follow this split structure for better semantics).

Also, it seems like the `Resource_hourly_matching.csv` file in the example case might be redundant. Is that correct?

Thanks again!

lbonaldo commented on src/write_outputs/hourly_matching_requirement/write_hourly_matching_revenue.jl:
```diff
@@ -0,0 +1,60 @@
+                              ((value.(EP[:vCHARGE_FLEX][FLEX,:]).data - value.(EP[:vP][FLEX, :])) * weighted_price)
+        end
+        if !isempty(ELECTROLYZERS)
+            temphmrev[ELECTROLYZERS] = hm.(gen[ELECTROLYZERS], tag = i) .*
```
Should we also include vP here?

wilson-ricks replied:
Not sure if this is referring to electrolyzers or flexible demand, but in either case no. Electrolyzers have vP=0 and for flex demand we don't have to correct an opposite-sign vP as in the model constraints

wilson-ricks committed:
Miscellaneous requested changes

Response to comments on PR#855. Added additional documentation and removed references to previous hydrogen hourly matching formulation.

lbonaldo committed:
Merge remote-tracking branch 'origin/develop' into hourlymatching

lbonaldo committed:
Fix docstrings

lbonaldo reviewed (approved):
I'm sorry for my late review. This looks good to me. And thanks so much again for this addition @wilson-ricks!
